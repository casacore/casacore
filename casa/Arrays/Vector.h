//# Vector.h: A 1-D Specialization of the Array Class
//# Copyright (C) 1993,1994,1995,1996,1998,1999,2000,2001,2002,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_VECTOR_H
#define CASA_VECTOR_H

//# Includes
#include <casacore/casa/Arrays/Array.h>

//# Forward declarations
//template <class T, class U> class vector; 
#if defined(WHATEVER_VECTOR_FORWARD_DEC)
WHATEVER_VECTOR_FORWARD_DEC;
#else
#include <casacore/casa/stdvector.h>
#endif

namespace casa { //#Begin namespace casa

// <summary> A 1-D Specialization of the Array class </summary>
// <reviewed reviewer="UNKNOWN" date="before2004/08/25" tests="" demos="">
// </reviewed>
//
// Vector objects are one-dimensional specializations (e.g., more convenient
// and efficient indexing) of the general Array class. You might also want
// to look at the Array documentation to see inherited functionality.
//
// Generally the member functions of Array are also available in
// Vector versions
// which take an integer where the array needs an IPosition. Since the Vector
// is one-dimensional, the IPositions are overkill, although you may
// use those versions if you want to.
// <srcblock>
// Vector<Int> vi(100);  // Vector 100 elements long.
// vi.resize(50);        // Now only 50 long.
// </srcblock>
//
// Slices may be taken with the Slice class. To take a slice, one "indexes" 
// with Slice(start, length, inc) where end and inc are optional.
// <srcblock>
// Vector<Float> vf(100);
// //...
// vf(Slice(0,50,2)) = vf(Slice(1,50,2));  // Copy values from odd onto even
// Vector<Float> firstHalf, secondHalf;
// firstHalf.reference(vf(Slice(0,50)));
// secondHalf.reference(vf(Slice(50,50)));
// // Now we have aliases for two slices into the Vector
// </srcblock>
//
// Element-by-element arithmetic and logical operations are available (in
// aips/ArrayMath.h and aips/ArrayLogical.h) as well as dot and cross
// products (in aips/MatrixMath.h).
//
// A Vector can be constructed from an STL <src>vector</src>. The reverse
// operation (<src>Array::tovector()</src>) can construct an STL
// <src>vector</src> from any Array.
// <note role=tip> To create any other STL container from an Array (or
// the reverse), always create from/to a <src>vector</src>, and use the
// range constructor to create from/to others (like set, list, deque). </note>
// 
// As with the Arrays, if the preprocessor symbol AIPS_DEBUG is
// defined at compile time invariants will be checked on entry to most
// member functions. Additionally, if AIPS_ARRAY_INDEX_CHECK is defined
// index operations will be bounds-checked. Neither of these should
// be defined for production code.

template<class T> class Vector : public Array<T>
{
public:
    // A zero-length Vector.
    Vector();

    // A Vector with a defined length and origin of zero.
    // <group>
    explicit Vector(uInt Length);
    explicit Vector(const IPosition& Length);
    // </group>

    // A Vector with a defined length and origin of zero.
    // Fill it with the initial value.
    // <group>
    Vector(uInt Length, const T &initialValue);
    Vector(const IPosition& Length, const T &initialValue);
    // </group>

    // Create a Vector from the given Block "other." Make it length "nr"
    // and copy over that many elements.
    Vector(const Block<T> &other, Int nr);
    // Create a Vector of length other.nelements() and copy over its values.
    explicit Vector(const Block<T> &other);

    // Create a reference to other.
    Vector(const Vector<T> &other);
    
    // Create a reference to the other array.
    // It is always possible if the array has zero or one axes.
    // If it has > 1 axes, it is only possible if the array has at most
    // one axis with length > 1. In that case the degenerated axes are removed.
    Vector(const Array<T> &other);

    // Create an Vector of a given shape from a pointer.
    Vector(const IPosition &shape, T *storage, StorageInitPolicy policy = COPY);
    // Create an Vector of a given shape from a pointer. Because the pointer
    // is const, a copy is always made.
    Vector(const IPosition &shape, const T *storage);

    // Create a Vector from an STL vector (see <src>tovector()</src> in
    // <linkto class=Array>Array</linkto>  for the reverse operation).
    // <note role=tip> Both this constructor and the tovector() are
    // defined in <src>Vector2.cc</src>. </note>
    template <class U>
    Vector(const vector<T, U> &other);

    // Create a Vector from a container iterator and its length.
    // <note> The length is used instead of last, because the distance
    // function needed to calculate the length can be expensive.
    // <br>A third dummy argument is unfortunately needed to avoid ambiguity
    // with another Vector constructor (taking two uInts).
    // </note>
    template<typename Iterator>
    Vector(Iterator first, size_t size, int dummy);

    // Define a destructor, otherwise the compiler makes a static one.
    virtual ~Vector();

    // Assign the other array (which must be of dimension one) to this vector.
    // If the shapes mismatch, this array is resized.
    virtual void assign (const Array<T>& other);

    // Create a reference to "other", which must be of dimension one.
    virtual void reference(const Array<T> &other);

    // Resize this Vector to the given length.
    // The default copyValues flag is False.
    //# Note that the 3rd resize function is necessary, because that
    //# function is virtual in the base class (otherwise it would
    //# be hidden).
    // Resize without argument is equal to resize(0, False).
    // <group>
    void resize(uInt len, Bool copyValues=False)
      { if (len != this->nelements()) resize (IPosition(1,len), copyValues); }
    virtual void resize(const IPosition &len, Bool copyValues=False);
    virtual void resize();
    // </group>

    // Assign to this Vector. If this Vector is zero-length, then resize
    // to be the same size as other. Otherwise this and other have to be
    // conformant (same size).
    // <br>Note that the assign function can be used to assign a
    // non-conforming vector.
    // <group>
    Vector<T> &operator=(const Vector<T> &other);
    // Other must be a 1-dimensional array.
    virtual Array<T> &operator=(const Array<T> &other);
    // </group>

    // Set every element of this Vector to Val.
    Array<T> &operator=(const T &val)
      { return Array<T>::operator=(val); }

    // Copy to this those values in marray whose corresponding elements
    // in marray's mask are True.
    Vector<T> &operator= (const MaskedArray<T> &marray)
      { Array<T> (*this) = marray; return *this; }

    // Convert a Vector to a Block, resizing the block and copying values.
    // This is done this way to avoid having the simpler Block class 
    // containing dependencies on the Vector class.
    void toBlock(Block<T> &other) const;

    // Single-pixel addressing. If AIPS_ARRAY_INDEX_CHECK is defined,
    // bounds checking is performed (not for [])..
    // <group>
    T &operator[](uInt index)
      { return (this->contiguous_p  ?  this->begin_p[index] : this->begin_p[index*this->inc_p(0)]); }
    const T &operator[](uInt index) const
      { return (this->contiguous_p  ?  this->begin_p[index] : this->begin_p[index*this->inc_p(0)]); }
    T &operator()(const IPosition &i)
      { return Array<T>::operator()(i); }
    const T &operator()(const IPosition &i) const 
      { return Array<T>::operator()(i); }
    T &operator()(uInt index)
      {
#if defined(AIPS_ARRAY_INDEX_CHECK)
	this->validateIndex(index);   //# Throws an exception on failure
#endif
        return *(this->begin_p + index*this->inc_p(0));
      }

    const T &operator()(uInt index) const
      {
#if defined(AIPS_ARRAY_INDEX_CHECK)
	this->validateIndex(index);   //# Throws an exception on failure
#endif
        return *(this->begin_p + index*this->inc_p(0));
      }
    // </group>

    // Take a slice of this vector. Slices are always indexed starting
    // at zero. This uses reference semantics, i.e. changing a value
    // in the slice changes the original.
    // <srcblock>
    // Vector<Double> vd(100);
    // //...
    // vd(Slice(0,10)) = -1.0; // First 10 elements of vd set to -1
    // </srcblock>
    // <group>
    Vector<T> operator()(const Slice &slice);
    const Vector<T> operator()(const Slice &slice) const;
    // </group>

    // Slice using IPositions. Required to be defined, otherwise the base
    // class versions are hidden.
    // <group>
    Array<T> operator()(const IPosition &blc, const IPosition &trc,
			const IPosition &incr)
      { return Array<T>::operator()(blc,trc,incr); }
    const Array<T> operator()(const IPosition &blc, const IPosition &trc,
                              const IPosition &incr) const
      { return Array<T>::operator()(blc,trc,incr); }
    Array<T> operator()(const IPosition &blc, const IPosition &trc)
      { return Array<T>::operator()(blc,trc); }
    const Array<T> operator()(const IPosition &blc, const IPosition &trc) const
      { return Array<T>::operator()(blc,trc); }
    Array<T> operator()(const Slicer& slicer)
      { return Array<T>::operator()(slicer); }
    const Array<T> operator()(const Slicer& slicer) const
      { return Array<T>::operator()(slicer); }
    // </group>

    // The array is masked by the input LogicalArray.
    // This mask must conform to the array.
    // <group>

    // Return a MaskedArray.
    MaskedArray<T> operator() (const LogicalArray &mask) const
      { return Array<T>::operator() (mask); }

    // Return a MaskedArray.
    MaskedArray<T> operator() (const LogicalArray &mask)
      { return Array<T>::operator() (mask); }

    // </group>


    // The array is masked by the input MaskedLogicalArray.
    // The mask is effectively the AND of the internal LogicalArray
    // and the internal mask of the MaskedLogicalArray.
    // The MaskedLogicalArray must conform to the array.
    // <group>

    // Return a MaskedArray.
    MaskedArray<T> operator() (const MaskedLogicalArray &mask) const
      { return Array<T>::operator() (mask); }

    // Return a MaskedArray.
    MaskedArray<T> operator() (const MaskedLogicalArray &mask)
      { return Array<T>::operator() (mask); }

    // </group>


    // The length of the Vector.
    // <group>
    void shape(Int &Shape) const
      { Shape = this->length_p(0); }
    const IPosition &shape() const
      { return this->length_p; }
    // </group>

    // Replace the data values with those in the pointer <src>storage</src>.
    // The results are undefined is storage does not point at nelements() or
    // more data elements. After takeStorage() is called, <src>unique()</src>
    // is True.
    // <group>
    virtual void takeStorage(const IPosition &shape, T *storage,
			     StorageInitPolicy policy = COPY);
    // Since the pointer is const, a copy is always taken.
    virtual void takeStorage(const IPosition &shape, const T *storage);
    // </group>

    // Verify that dimensionality is 1 and then call Array<T>::ok()
    virtual Bool ok() const;

protected:
    // Remove the degenerate axes from other and store result in this vector.
    // An exception is thrown if removing degenerate axes does not result
    // in a vector.
    virtual void doNonDegenerate(const Array<T> &other,
                                 const IPosition &ignoreAxes);

private:
    // Helper functions for constructors.
    void initVector(const Block<T> &, Int nr);      // copy semantics
};

} //#End namespace casa

#ifndef CASACORE_NO_AUTO_TEMPLATES
#include <casacore/casa/Arrays/Vector.tcc>
#include <casacore/casa/Arrays/Vector2.tcc>
#endif //# CASACORE_NO_AUTO_TEMPLATES
#endif
