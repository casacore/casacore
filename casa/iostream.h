//# iostream.h: Interim solution for standard/nonstandard system iostream
//# Copyright (C) 2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_IOSTREAM_H
#define CASA_IOSTREAM_H

//# Define the C standard C++ include file. 
//# This is an interim solution to cater for the SGI non-existence of
//# them (e.g. <cstring>)

//# Make sure any special macros are set
#include <casacore/casa/aips.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

using std::istream;
using std::ostream;
using std::iostream;
using std::cin;
using std::cout;
using std::cerr;
using std::endl;
using std::flush;
using std::ws;

using std::ios;
using std::streampos;
using std::streamoff;
using std::streamsize;
using std::dec;
using std::hex;
using std::oct;

using std::internal;
using std::left;
using std::right;
using std::fixed;
using std::scientific;

using std::boolalpha;
using std::noboolalpha;
using std::showbase;
using std::noshowbase;
using std::showpoint;
using std::noshowpoint;
using std::showpos;
using std::noshowpos;
using std::skipws;
using std::noskipws;
using std::uppercase;
using std::nouppercase;

} //# NAMESPACE CASACORE - END

#endif
