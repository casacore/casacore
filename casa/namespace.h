#ifndef NAMESPACE_H
#define NAMESPACE_H 
using namespace casa;
#endif
