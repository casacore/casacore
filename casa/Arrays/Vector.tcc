//# Vector.cc: A 1-D Specialization of the Array Class
//# Copyright (C) 1993,1994,1995,1996,1997,1998,1999,2000,2001,2002,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/Arrays/Vector.h>
#include <casa/Arrays/ArrayError.h>
#include <casa/Arrays/Slice.h>
#include <casa/Arrays/MaskedArray.h>
#include <casa/BasicMath/Math.h>
#include <casa/Utilities/Copy.h>
#include <casa/Utilities/Assert.h>
#include <casa/iostream.h>

namespace casa { //#Begin casa namespace

template<class T> Vector<T>::Vector()
  : Array<T>(IPosition(1,0))
{
    DebugAssert(ok(), ArrayError);
}


template<class T> Vector<T>::Vector(uInt Length)
: Array<T>(IPosition(1, Length))
{
    DebugAssert(ok(), ArrayError);
}

template<class T> Vector<T>::Vector(const IPosition& len)
  : Array<T>(len)
{
    if (len.nelements() != 1) this->throwNdimVector();
}

template<class T> Vector<T>::Vector(uInt Length, const T &initialValue)
: Array<T>(IPosition(1, Length), initialValue)
{
    DebugAssert(ok(), ArrayError);
}

template<class T> Vector<T>::Vector(const IPosition& len, const T &initialValue)
  : Array<T>(len, initialValue)
{
    if (len.nelements() != 1) this->throwNdimVector();
}

template<class T> Vector<T>::Vector(const Block<T> &other, Int nr)
: Array<T>(IPosition(1, other.nelements()))
{
    initVector (other, nr);
    DebugAssert(ok(), ArrayError);
}

template<class T> Vector<T>::Vector(const Block<T> &other)
: Array<T>(IPosition(1, other.nelements()))
{
    initVector (other, 0);
    DebugAssert(ok(), ArrayError);
}

// Copy from the block. Copy the number of elements specified or
// the number of elements in the block if nr <= 0.
// <thrown>
//    <item> ArrayError
// </thrown>
template<class T> void Vector<T>::initVector(const Block<T> &other, Int nr)
{
    uInt n = nr;
    if (nr <= 0) {
	n = other.nelements();
    }
    if (n > other.nelements())
	throw(ArrayError("Vector<T>::initVector(const Block<T> &other"
				   ", Int nr) - nr > other.nelements()"));
    if (this->nelements() != n) {
	this->resize(n);
    }
    for (uInt i=0; i < n; i++) {
	this->begin_p[i] = other[i];
    }
    return;
}

template<class T> Vector<T>::Vector(const Vector<T> &other)
: Array<T>(other)
{
    DebugAssert(ok(), ArrayError);
}

// <thrown>
//    <item> ArrayNDimError
// </thrown>
template<class T> Vector<T>::Vector(const Array<T> &other)
: Array<T>(other)
{
    // If not 1 dimension, adjust shape if possible.
    if (this->ndim() != 1) {
        this->checkVectorShape();
    }
    DebugAssert(ok(), ArrayError);
}

template<class T>
Vector<T>::Vector(const IPosition &shape, T *storage, 
		  StorageInitPolicy policy)
  : Array<T>(shape, storage, policy)
{
    if (shape.nelements() != 1) this->throwNdimVector();
}

template<class T>
Vector<T>::Vector(const IPosition &shape, const T *storage)
  : Array<T>(shape, storage)
{
    if (shape.nelements() != 1) this->throwNdimVector();
}

template<class T> Vector<T>::~Vector()
{
    // Nothing
}


template<class T> void Vector<T>::resize()
{
    resize (IPosition(1,0), False);
}

// <thrown>
//    <item> ArrayConformanceError
// </thrown>
template<class T> void Vector<T>::resize(const IPosition &l, Bool copyValues)
{
    DebugAssert(ok(), ArrayError);
    if (l.nelements() != 1) this->throwNdimVector();
    if (copyValues) {
        Vector<T> oldref(*this);
	Array<T>::resize(l);
	size_t minNels = std::min(this->nelements(), oldref.nelements());
	objcopy(this->begin_p, oldref.begin_p, minNels,
		uInt(this->inc_p(0)), uInt(oldref.inc_p(0)));
    } else {
	Array<T>::resize(l);
    }
    DebugAssert(ok(), ArrayError);
}


// <thrown>
//    <item> ArrayError
// </thrown>
template<class T> void Vector<T>::assign (const Array<T>& other)
{
    DebugAssert(ok(), ArrayError);
    if (other.ndim() != 1) this->throwNdimVector();
    Array<T>::assign (other);
}

// <thrown>
//    <item> ArrayError
// </thrown>
template<class T> void Vector<T>::reference(const Array<T> &other)
{
    DebugAssert(ok(), ArrayError);
    if (other.ndim() != 1) this->throwNdimVector();
    Array<T>::reference(other);
}

template<class T> Vector<T> &Vector<T>::operator=(const Vector<T> &other)
{
    DebugAssert(ok(), ArrayError);
    if (this != &other) {
        if (! this->copyVectorHelper (other)) {
	    // Block was empty, so allocate new block.
	    this->data_p  = new Block<T> (this->length_p(0));
	    this->begin_p = this->data_p->storage();
	}
	this->setEndIter();
	objcopy (this->begin_p, other.begin_p, this->nels_p,
		 this->inc_p(0), other.inc_p(0));
    }
    return *this;
}

// Copy a vector to a block. 
template<class T> void Vector<T>::toBlock(Block<T> & other) const
{
    DebugAssert(ok(), ArrayError);
    uInt vec_length = this->nelements();
    // Make sure the block has enough space, but there is no need to copy elements
    other.resize(vec_length, True, False);
    objcopy(other.storage(), this->begin_p, this->nels_p, 1U, this->inc_p(0));
}

template<class T> Array<T> &Vector<T>::operator=(const Array<T> &a)
{
    DebugAssert(ok(), ArrayError);
    Vector<T> tmp(a);
    (*this) = tmp;
    return *this;
}

// <thrown>
//    <item> ArrayError
// </thrown>
template<class T> Vector<T> Vector<T>::operator()(const Slice &slice)
{
    DebugAssert(ok(), ArrayError);
    Int b, l, s;       // begin length step
    if (slice.all()) {
	b = 0;
	l = this->length_p(0);
	s = 1;
    } else {
	b = slice.start();
	l = slice.length();
	s = slice.inc();
    }

    // Check that the selected slice is valid
    if (s < 1) {
	throw(ArrayError("Vector<T>::operator()(Slice) : step < 1"));
    } else if (l < 0) {
	throw(ArrayError("Vector<T>::operator()(Slice) : length < 0"));
    } else if (b+(l-1)*s >= this->length_p(0)) {
	throw(ArrayError("Vector<T>::operator()(Slice) : Desired slice extends"
			 " beyond the end of the array"));
    } else if (b < 0) {
	throw(ArrayError("Vector<T>::operator()(Slice) : start of slice before "
			 "beginning of vector"));
    }

    // Create the slice. This could also be done with the Array<T>::operator()
    // slice functions, however it's simple for vectors, and this will be
    // more efficient.

    // Create the vector that will be the slice into this
    Vector<T> vp(*this);

    // Increment vp's begin so that it is at the selected position
    vp.begin_p += b*this->steps()[0];
    vp.inc_p(0) *= s;
    vp.length_p(0) = l;
    vp.nels_p = l;
    vp.contiguous_p = vp.isStorageContiguous();
    vp.makeSteps();

    return vp;
}

template<class T> const Vector<T> Vector<T>::operator()
  (const Slice &slice) const
{
    return const_cast<Vector<T>*>(this)->operator() (slice);
}

template<class T>
void Vector<T>::doNonDegenerate (const Array<T> &other,
                                 const IPosition &ignoreAxes)
{
    Array<T> tmp(*this);
    tmp.nonDegenerate (other, ignoreAxes);
    if (tmp.ndim() != 1) this->throwNdimVector();
    reference (tmp);
}

template<class T>
void Vector<T>::takeStorage(const IPosition &shape, T *storage,
			    StorageInitPolicy policy)
{
    if (shape.nelements() != 1) this->throwNdimVector();
    Array<T>::takeStorage(shape, storage, policy);
}

template<class T>
void Vector<T>::takeStorage(const IPosition &shape, const T *storage)
{
    if (shape.nelements() != 1) this->throwNdimVector();
    Array<T>::takeStorage(shape, storage);
}

template<class T> Bool Vector<T>::ok() const
{
    return  this->ndim() == 1  &&  Array<T>::ok();
}

} //#End casa namespace
