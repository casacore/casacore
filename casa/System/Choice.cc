//# Choice.cc: Ask a choice to the user
//# Copyright (C) 2004
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#include <casacore/casa/System/Choice.h>
#include <casacore/casa/Arrays/Vector.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

//# Default is no choice function, thus return first choice.
Choice::ChoiceFunc* Choice::theirChoiceFunc = 0;


String Choice::choice (const String& descriptiveText,
		       const Vector<String>& choices)
{
  if (choices.nelements() == 0) {
    return "";
  }
  if (theirChoiceFunc == 0) {
    return choices[0];
  }
  return theirChoiceFunc (descriptiveText, choices);
}

Choice::ChoiceFunc* Choice::setChoiceFunc (Choice::ChoiceFunc* func)
{
  Choice::ChoiceFunc* tmp = theirChoiceFunc;
  theirChoiceFunc = func;
  return tmp;
}

String Choice::ostreamChoice (std::ostream& os,
			      const String& descriptiveText,
			      const Vector<String>& choices)
{
  if (choices.nelements() == 0) {
    return "";
  }
  char answer[256];
  while (True) {
    os << descriptiveText << " ([" << choices(0) << ']';
    for (uInt i=1; i<choices.nelements(); i++) {
      os << ',' << choices[i];
    }
    os << "): ";
    cin.getline (answer, sizeof(answer));
    String str(answer);
    if (str.size() == 0) {
      return choices[0];
    }
    for (uInt i=0; i<choices.nelements(); i++) {
      if (str == choices[i]) {
	return choices[i];
      }
    }
    os << "'" << str << "' is an invalid answer; retry" << std::endl;
  }
  return "";
}

} //# NAMESPACE CASACORE - END

