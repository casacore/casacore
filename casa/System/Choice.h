//# Choice.h: Ask a choice to the user
//# Copyright (C) 2004
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#ifndef CASA_CHOICE_H
#define CASA_CHOICE_H


//# Includes
#include <casacore/casa/aips.h>
#include <casacore/casa/BasicSL/String.h>
#include <iostream>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

//# Forward Declarations
template<class T> class Vector;

// <summary>
// Class to ask a user a choice
// </summary>

// <reviewed reviewer="" date="yyyy/mm/dd" tests="" demos="">
// </reviewed>

// <prerequisite>
// </prerequisite>

// <etymology>
// </etymology>

// <synopsis>
// </synopsis>

class Choice
{
public:
  // Define the signature of the choice function.
  typedef String ChoiceFunc (const String& descriptiveText,
			     const Vector<String>& choices);

  // Get a choice from the user.
  // The choice function to be used can be set using setChoiceFunc.
  // It can, for instance, be done by ObjectController.
  // Initially no choice function is set. In that case it returns
  // the first choice (so that should be the default choice).
  // If choices is zero length, an empty string is returned.
  static String choice (const String& descriptiveText,
			const Vector<String>& choices);

  // Set the choice function.
  // It returns the old choice function.
  static ChoiceFunc* setChoiceFunc (ChoiceFunc* func);

  // A choice function asking on stderr.
  static String stderrChoice (const String& descriptiveText,
			      const Vector<String>& choices)
    { return ostreamChoice (std::cerr, descriptiveText, choices); }

  // A choice function asking on stdout.
  // It outputs the descriptiveText followed by a blank, the options (enclosed
  // in parentheses) and a colon.
  // The default option is shown in square brackets.
  static String stdoutChoice (const String& descriptiveText,
			      const Vector<String>& choices)
    { return ostreamChoice (std::cout, descriptiveText, choices); }


private:
  // Ask on an ostream.
  static String ostreamChoice (std::ostream&,
			       const String& descriptiveText,
			       const Vector<String>& choices);

  //# Pointer to the choice function.
  static ChoiceFunc* theirChoiceFunc; 

};



} //# NAMESPACE CASACORE - END

#endif
