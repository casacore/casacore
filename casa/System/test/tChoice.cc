//# tChoice.cc: Test program for class Choice
//# Copyright (C) 2004
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//#
//# $Id$

#include <casa/System/Choice.h>
#include <casa/Arrays/Vector.h>
#include <casa/Utilities/Assert.h>
#include <iostream>

#include <casa/namespace.h>
int main()
{
  Vector<String> choices(3);
  choices[0] = "no";
  choices[1] = "yes";
  choices[2] = "none";

  // The default is no.
  AlwaysAssertExit (Choice::choice ("ask", choices) == "no");

  // Set the choice function to asking on stdout.
  Choice::setChoiceFunc (Choice::stdoutChoice);

  // Ask the choice 3 times and show the result.
  std::cout << Choice::choice ("Give choice1", choices) << std::endl;
  std::cout << Choice::choice ("Give choice2", choices) << std::endl;
  std::cout << Choice::choice ("Give choice3", choices) << std::endl;

  return 0;
}
