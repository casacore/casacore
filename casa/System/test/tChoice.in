no
YES
yes
non
no1
none
