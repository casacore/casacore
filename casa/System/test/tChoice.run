$casa_checktool ./tChoice < tChoice.in
