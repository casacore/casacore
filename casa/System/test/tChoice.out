Give choice1 ([no],yes,none): no
Give choice2 ([no],yes,none): 'YES' is an invalid answer; retry
Give choice2 ([no],yes,none): yes
Give choice3 ([no],yes,none): 'non' is an invalid answer; retry
Give choice3 ([no],yes,none): 'no1' is an invalid answer; retry
Give choice3 ([no],yes,none): none
